### PR TITLE
Allow commenting and resolving together

### PR DESCRIPTION
**Pull request by benarmston**
This PR adds support for commenting and resolving with one submission.  It has the same UI/UX as reopening cases. That is a dialog is shown in which the comment can be added.

I have some WIP that changes the UI/UX to add a "Comment and resolve" button next to the standard form submission button.  It isn't quite ready yet, so I've extracted what I have immediately ready here.

This PR is probably most easily reviewed after #762 is merged.

Refs #709

### CODE REVIEW

**Comment by jamesremuscat** [comment]
If you're not the assigned engineer for a case, then you are unable to add a comment, but the resolve-case form still prompts you to enter one; if you do so, the comment gets swallowed and disappears.

We could either:
1. Hide the comment box in the resolve case form if `policy(case).comment?` is `false`
2. Allow any engineer to comment when resolving a case
3. Only allow the assigned engineer to resolve a case

For now I'm inclined to opt for 1. since it's closest to our existing behaviour.

File: app/views/cases/_case_state_controls.html.erb [hunk 1]
```diff
@@ -1,16 +1,67 @@
...
+          <div class="modal-content">
+            <%= form_tag resolve_case_path(kase), method: :post do |f| %>
+              <div class="modal-header">
+                <h5 class="modal-title" id="<%= modal_label_id %>">
+                  Resolve case
+                </h5>
+                <button class="close" data-dismiss="modal">×</button>
+              </div>
+              <div class="modal-body">
+                <p>Comment</p>
```
Comment (jamesremuscat) [inline app/views/cases/_case_state_controls.html.erb]: We should make it clear that the comment is optional.

**Comment by benarmston** [comment]
The above two commits fix both of the issues you have raised.  The UI/UX has changed so that there is now a "resolve" button next to the standard "submit" button.

This should make it obvious that the comment is optional, as the form can generally be submitted without a comment.

As the comment form is only shown for engineers assigned to the case, we no longer prompt for a comment that will be ignored.

This new method does expose a new potential issue.  Someone could carefully craft a request to add a comment, change some fields and resolve a case.  If the case cannot be resolved, but the request to change the fields and add a comment is legitimate, we change the fields and add the comment, but silently ignore the attempt to resolve the case.

This doesn't expose any attack vectors, and the UI only displaying the resolve button when relevant prevents this issue from being uncovered by our users.  So its probably more of an API oddity that might cause us some head scratching at some point.

---

In terms of the UI I think this is what we should be aiming for.  The UX of sending a plethora of emails still exists, but that is looking much more easily resolved now than it did before.

**Comment by jamesremuscat** [comment]
This is looking really good - completely agree that this is the UI we should be aiming for, and the email UX will follow.

selection_013

I wonder if the first button should be labelled "Update and resolve case" to make it clearer that its actions are a superset of "Update case details"? @mjtko?

> Someone could carefully craft a request to add a comment, change some fields and resolve a case. If the case cannot be resolved, but the request to change the fields and add a comment is legitimate, we change the fields and add the comment, but silently ignore the attempt to resolve the case.

I don't think we need to worry about this too greatly - as you say, it's not an attack vector, and our UI prevents it from occurring in normal usage. We might consider adding a flash message if we do end up getting confused by it. (An interesting case to consider might be where updating the case attributes renders it unresolvable by e.g. unsetting `time_worked` - that might legitimately occur through the UI, and we actually handle that situation quite well with an error flash as things stand.)

**Comment by benarmston** [comment]
@jamesremuscat your comment about unsetting `time_worked` when resolving has led me to realise that this PR doesn't permit updating the time worked (to a sensible value) and resolving in the same request.  I have a fix and tests for this.  Let me know if you want them including on this PR or a separate one.

**Comment by jamesremuscat** [comment]
@benarmston The above is my only outstanding comment on this PR; once resolved I'll be very happy to merge it.

**Pull request merged**